teps/0056-pipelines-in-pipelines.md

```diff
@@ -352,6 +363,451 @@ main-`PipelineRun` to be able to fetch and access the sub-`PipelineRun`'s full s
+  `PipelineRun` or `TaskRun`), users should be able to control execution as needed by their context without having to
+  modify `Tasks` and `Pipeline`.
+
+### Specification - Reorganize `PipelineTask`
+
+We could reorganize the `PipelineTask` to better support scaling as we add more types that can be specified and 
+executed in a `Pipeline`, beyond `Tasks` and `Custom Tasks`. A possible solution to this could be adding new fields
+under `PipelineTask` for `Ref` and `Spec`, which would then contain fields specific for the supported `PipelineTask`
+types, as follows:
+
+```go
+type PipelineTask struct {
+    Name     string                `json:"name,omitempty"`
+    TaskRef  *TaskRef              `json:"taskRef,omitempty"`
+    TaskSpec *EmbeddedTask         `json:"taskSpec,omitempty"`
+    Ref      *PipelineTaskRef      `json:"ref,omitempty"`
+    Spec     *PipelineTaskSpec     `json:"spec,omitempty"`
+    ...
+}
+
+type PipelineTaskRef struct {
+	Task       *TaskRef             `json:"task,omitempty"`
+	CustomTask *TaskRef             `json:"customTask,omitempty"`
+	Pipeline   *PipelineRef         `json:"pipeline,omitempty"`
+}
+
+type PipelineTaskSpec struct {
+	Task        *TaskSpec               `json:"task,omitempty"`
+	CustomTask  *CustomTaskSpec         `json:"customTask,omitempty"`
+	Pipeline    *PipelineSpec           `json:"pipeline,omitempty"`
+	Metadata    PipelineTaskMetadata    `json:"metadata,omitempty"`
+}
+
+type CustomTaskSpec struct {
+	runtime.TypeMeta          `json:",inline,omitempty"`
+	Spec runtime.RawExtension `json:"spec,omitempty"`
+}
+```
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: clone-scan-notify
+spec:
+  tasks:
+    - name: git-clone
+      ref:
+        task:
+          name: git-clone
+    - name: security-scans
+      ref:
+        pipeline:
+          name: security-scans
+    - name: notification
+      taskRef: # for backwards compatibility as migration happens
+        name: notification
+```
+
+We could support the existing `TaskRef` and `TaskSpec` fields would continue for the foreseeable future, but anything
+specified under the new `Ref` or `Spec` fields would take priority. Today, we use `EmbeddedTask`, which has `TaskSpec`
+inlined for `Tasks`, `runtime.TypeMeta` and `runtime.RawExtension` for custom tasks, as well as the `Metadata` field.
+Instead, we could create a new `CustomTaskSpec` type, and just use `TaskSpec` for `Tasks` and `PipelineSpec` for 
+`Pipelines`, while adding the `Metadata` field to `PipelineTaskSpec`.
+
+However, this approach does not eliminate the problem of handling `PipelineRunSpec` fields. In addition, this is a 
+big change to the `PipelineTask` API that can be pursued separately from supporting `Pipelines` in `Pipelines`, one 
+does not block the other. We propose keeping the scope of this TEP focused on supporting `Pipelines` in `Pipelines` by 
+decoupling the reorganization of `PipelineTask`, which remains an option to pursue later.
```

Comment:
Although I tend to agree, if we "ever" want to go that way (re-org the *PipelineTask*), it would be better to do it prior to pipeline-in-pipeline than after, as it would had "one more thing to migrate to / from". pipeline-in-pipeline could even be the first thing supported by that new approach

Reply:
expounded on the reasons not to do this alternative, please take a look 🙏🏾